src/psyclone/tests/psyir/frontend/fparser2_derived_type_test.py

```diff
@@ -176,11 +178,163 @@ def test_derived_type_accessibility():
                                  "  integer :: flag\n"
                                  "  real, public :: scale\n"
                                  "end type my_type\n")
-    fparser2spec = Specification_Part(reader)
+    fparser2spec = Fortran2003.Specification_Part(reader)
     processor.process_declarations(fake_parent, fparser2spec.content, [])
     sym = symtab.lookup("my_type")
     assert isinstance(sym, TypeSymbol)
     flag = sym.datatype.lookup("flag")
     assert flag.visibility == Symbol.Visibility.PRIVATE
     scale = sym.datatype.lookup("scale")
     assert scale.visibility == Symbol.Visibility.PUBLIC
+
+
+def test_derived_type_ref(f2008_parser):
```

Comment:
These are both nice tests. Would it be possible to add a "member" property in structure_referece.py that returns "children[0]" for the semantic navigations in this tests (e.g `assing.lhs.member.member`)

There will still be a semantic navigation property missing for the array nodes, but this is a bit more difficult because it can be 1 or more nodes (maybe always return a list) and they are represented by different children (will need to be implemented in multiple nodes). But is something to consider.

Reply:
That's a good idea. As you suggest, I'm also attempting to add an `indices` property to the array nodes but this does lead to a lot of duplication - `ArrayMixin`, `ArrayOfStructuresReference` and `ArrayOfStructuresMember` all need implementations. I could avoid this by having the implementation in ArrayMixin check on `type(self)` and thus decide whether or not to include the first child in the list of returned nodes. What do you think?

Reply:
Perhaps we should have an `ArrayOfStructures` abstract base class?

Reply:
To be honest I think I would live with the duplication here, which is just twice for each case: ArrayReference and ArrayMember return children starting by number 0 and ArrayOfStructuresReference and ArrayOfStructuresMember return children starting by number 1.

The reason I am not convinced of implementing it on ArrayMixing checking type(self) is that this seems to convey the same information available in `_children_valid_format` and the `_validate_child` method (which child is what). And since it is the same information maybe better to encapsulate in the same place.

What do you think?

Reply:
I've gone for a new `ArrayOfStructures` mixin class which provides appropriately specialised `_validate_child` and `indices` methods. That way we have pretty much no code duplication.

Reply:
Thanks for doing the extra work, I like how it looks!